src/Processors/Formats/Impl/Parquet/ParquetBloomFilterCondition.cpp

```diff
@@ -0,0 +1,295 @@
+        uint8_t buffer[buffer_size] = {0};
+
+        auto column_size = data_column->size();
+
+        auto hashes_column = ColumnUInt64::create(column_size);
+        ColumnUInt64::Container & hashes_internal_data = hashes_column->getData();
+
+        for (size_t i = 0; i < column_size; ++i)
+        {
+            const auto data_view = data_column->getDataAt(i).toView();
+
+            const auto ba = createByteArray(data_view, data_column->getDataType(), buffer, buffer_size);
+
+            const auto hash = bloom_filter->Hash(&ba);
+
+            hashes_internal_data[i] = hash;
+        }
+
+        return hashes_column;
+    }
+
+    bool maybeTrueOnBloomFilter(ColumnPtr hash_column, const std::unique_ptr<parquet::BloomFilter> & bloom_filter, bool match_all)
+    {
+        const auto * uint64_column = typeid_cast<const ColumnUInt64 *>(hash_column.get());
+
+        if (!uint64_column)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Hash column must be UInt64.");
+
+        const ColumnUInt64::Container & hashes = uint64_column->getData();
+
+        for (const auto hash : hashes)
+        {
+            bool found = bloom_filter->FindHash(hash);
+
+            if (match_all && !found)
+                return false;
+            if (!match_all && found)
+                return true;
+        }
+
+        return match_all;
+    }
+
+    DataTypePtr getPrimitiveType(const DataTypePtr & data_type)
+    {
+        if (const auto * array_type = typeid_cast<const DataTypeArray *>(data_type.get()))
+        {
+            if (!typeid_cast<const DataTypeArray *>(array_type->getNestedType().get()))
+                return getPrimitiveType(array_type->getNestedType());
+            else
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unexpected type {} of bloom filter index.", data_type->getName());
+        }
+
+        if (const auto * nullable_type = typeid_cast<const DataTypeNullable *>(data_type.get()))
+            return getPrimitiveType(nullable_type->getNestedType());
+
+        if (const auto * low_cardinality_type = typeid_cast<const DataTypeLowCardinality *>(data_type.get()))
+            return getPrimitiveType(low_cardinality_type->getDictionaryType());
+
+        return data_type;
+    }
+}
+
+ParquetBloomFilterCondition::ParquetBloomFilterCondition(const std::vector<ConditionElement> & condition_)
+: condition(condition_)
+{
+}
+
+bool ParquetBloomFilterCondition::mayBeTrueOnRowGroup(const IndexColumnToColumnBF & column_index_to_column_bf) const
+{
+    using Function = ConditionElement::Function;
+    std::vector<BoolMask> rpn_stack;
+
+    for (const auto & element : condition)
+    {
+        if (element.function == Function::FUNCTION_EQUALS
+                 || element.function == Function::FUNCTION_NOT_EQUALS)
+        {
+            bool maybe_true = maybeTrueOnBloomFilter(element.columns[0], column_index_to_column_bf.at(element.key_columns[0]), false);
```

Comment:
If the first row group has bloom filter, but some other row group does not, this `at` will throw. Same for the other `maybeTrueOnBloomFilter` call.

Reply:
Is it possible for a bloom filter to be available on rg 0 and not on the others? I thought it existed in a column basis (i.e, if column X has bloom filter enabled, all row groups that contain that column will have bloom filters. You can't disable it on a per row group basis)

Reply:
`parquet.thrift` just says `optional i64 bloom_filter_offset` in column chunk; no comment promising that it's consistent across row groups, no bloom filter info in SchemaElement. It's easy to imagine a parquet encoder that e.g. omits bloom filter if there are too many distinct values, similar to how parquet encoders usually omit string min/max statistics if the string is too long. Idk whether any popular encoder implementations currently do that.